socket/socket.js

```diff
@@ -0,0 +1,284 @@
+            singleUser: false,
+            chatList: chatlistResponse,
+          });
+
+          // this emit updates the status of any new online or offline users to everyone
+          // (useful for indicating which of your contacts is online/offline)
+          socket.broadcast.emit(`chat-list-response`, {
+            error: false,
+            singleUser: true,
+            userInfo: UserInfoResponse,
+          });
+        } catch (error) {
+          io.to(socket.id).emit(`chat-list-response`, {
+            error: true,
+            chatList: [],
+          });
+        }
+      }
+    });
+
+    /**
+     * send the messages to the user
+     */
+    socket.on(`add-message`, async (data) => {
+      if (data.message === "") {
+        io.to(socket.id).emit(`add-message-response`, {
+          error: true,
+          message: "Message not found",
+        });
+      } else if (data.fromUserId === "") {
+        io.to(socket.id).emit(`add-message-response`, {
+          error: true,
+          message:
+            "server cannot process the request since requested fromUserId is missing from data",
+        });
+      } else if (data.toUserId === "") {
+        io.to(socket.id).emit(`add-message-response`, {
+          error: true,
+          message:
+            "server cannot process the request since requested toUserId is missing from data",
+        });
+      } else {
+        try {
+          const [toSocketId, messageResult] = await Promise.all([
+            getUserSocketId(data.toUserId),
+            insertMessages(data),
+          ]);
+          io.to(toSocketId).emit(`add-message-response`, {
+            error: false,
+            chatMsg: messageResult.chatMsg,
+          });
+        } catch (error) {
+          io.to(socket.id).emit(`add-message-response`, {
+            error: true,
+            message:
+              "Failed to add messages to the db or retrieve user socket info.",
+          });
+        }
+      }
+    });
+
+    /**
+     * Set user to offline (logout of socket)
+     */
+    socket.on("logout", async (data) => {
+      try {
+        const userId = data.userId;
+        await logout(userId);
+        const userInfo = await getUserInfo(userId);
+        io.to(socket.id).emit(`logout-response`, {
+          error: false,
+          message: "User is now offline",
+          userId,
+        });
+        // broadcast to all users who have this user in their chat list that
+        // the user has become offline
+        socket.broadcast.emit(`chat-list-response`, {
+          error: false,
+          userDisconnected: true,
+          userInfo,
+        });
+      } catch (error) {
+        io.to(socket.id).emit(`logout-response`, {
+          error: true,
+          message: "Failed to set user to offline",
+          userId,
+        });
+      }
+    });
+
+    /**
+     * sending the disconnected user to all socket users.
+     */
+    socket.on("disconnect", async () => {
```

Comment:
What is the difference between log out and disconnect?

Reply:
Nothing really, can remove this. :)